### ISSUE TITLE
Typings definitons for Typscript

### ISSUE THREAD

**Issue by lebsral** [olivernn/lunr.js#241]
I'm trying to use lunr in a fresh Angular 2 app installed with the cli.

In the app.module.ts I'm trying to import lunr with 

```TypeScript
import { lunr } from 'lunr';
```

But I keep getting errors about that not being a module, or it not finding it.  (yes lunr is in the node_modules folder)

I think/guess that it needs a Typings definition?  I found this one.  But it seems to be considerably out of date, and doesn't solve the problem
https://www.npmjs.com/package/@types/lunr

Or is that not what I need?

TIA

**Comment by olivernn** [comment]
I don't have any experience with either typescript or angular 2 so I'm not really sure what to suggest. If you do figure out what is needed please update this issue with what you found.

**Comment by olivernn** [comment]
Issue has gone stale, closing.

**Comment by dolphin1414** [comment]
Hello! I have the same problem.

**Comment by gblok** [comment]
const lunr = require('lunr')
lurn/lurn.js not exported modules, require as file

**Comment by ncphillips** [comment]
I would like to open this issue up again. 

We're using Lunr in a Typescript project and I may have some spare time to help. I would prefer to keep the typings in this project itself, for reasons including:

* Fewer dependencies
* Nice to keep definitions close to source code
* Projects using lunr don't have to worry about definition version not matching source version
* Versionin number makes more sense
* Easier for me to contribute to this project then to definitely typed.
* Remembering to keep up changes is easier

**Comment by olivernn** [comment]
I'm happy to re-open the issue as I think having TypeScript definitions would be awesome.

That said, I really don't think they should live in this repository.

I'm not familiar with the process of creating a typings file, can this not be something that is automated for each new release? Or some kind of test that can highlight missing or incorrect typings?

Do you have some examples of other JavaScript libraries and where they keep their type definitions?

**Comment by gnomeontherun** [comment]
Types exist https://www.npmjs.com/package/@types/lunr

**Comment by kchiragowni** [comment]
Hi @gnomeontherun,

Have you got an example or know docs website on how to implement Lunr.js search in Angular 2/4

Thanks

**Comment by gnomeontherun** [comment]
@kchiragowni There is nothing special to do to implement it with Angular 2+. Lunr is available on the global window, so you just access it there.

**Comment by seantanly** [comment]
It seems that the typings at https://www.npmjs.com/package/@types/lunr is quite outdated. For example for lunr v2.1, the index is immutable, but the typings provided still defines the mutation functions such as add/remove on the index.

**Comment by seantanly** [comment]
@olivernn `axios` is one successful library that maintains the typings within its project. The typings are declared in `index.d.ts` at root and declared in `package.json` as `"typings": "./index.d.ts"`.

https://github.com/mzabriskie/axios

With that its typescript users enjoy the benefits which @ncphillips has mentioned, not having to manually add typings from @types (DefinitelyTyped) and sometimes realize the typings are outdated and can't be relied upon.

**Comment by jbouzekri** [comment]
@kchiragowni  : As @gnomeontherun said, there is really nothing special to do.

If you are using angular cli : 

```
npm install --save lunr
npm install --save-dev @types/lunr
```
In tsconfig.json, add : 

```
{
  "compilerOptions": {
    "types":[
      "lunr"
    ]
  }
}
```

And in your component ts file, you can do the same thing as in the lunr guide : 

```
import * as lunr from 'lunr';

const idx = lunr(function () {
  this.ref('name')
  this.field('text')

  documents.forEach(function (doc) {
    this.add(doc)
  }, this)
});
```

**Comment by seantanly** [comment]
I have rewritten the typescript definitions for lunr based on v2.1.3 and submitted the pull request to DefinitelyTyped.

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19598

Once it's approved/merged and pushed to npm, typescript users should be able to install the latest definition with `npm install --save-dev @types/lunr`.

Still, it would be great if the definitions can live as part of this project, so that it gets updated as the project evolves.

**Comment by jbouzekri** [comment]
@seantanly : Will we still be able to install the definition for v1.0.0 using `npm install --save-dev @types/lunr` ?

**Comment by seantanly** [comment]
@jbouzekri Definitely, you simply need to define the version of the `@types/lunr` package you're going to install.

Looking at the current info,
```
npm info @types/lunr
```
The latest version is `0.5.29`which I assume is the definition that supports lunr v1.0.0. 

With that info, you just need to run `npm install --save-dev @types/lunr@~0.5.29`, which should peg your version of the package to `~0.5.29`.

The new package version of the type definition should be `2.1.x`.

**Comment by jbouzekri** [comment]
@seantanly : thanks. It is a little confusing how there is a single repository of typings that produces all these npm @types registrations with different versions.

**Comment by seantanly** [comment]
@jbouzekri DefinitelyTyped is a repository for all external typescript typings, and the repository is automated to publish npm packages under the scope `@types` upon successful PR merges. The version/author/misc info of the packages is determined by the magic comments inside `index.d.ts`. For example, the version that I have submitted indicated `2.1`, thus the published package version should be `2.1.x`.

https://github.com/seantanly/DefinitelyTyped/blob/84f0260d81d798231bab483740a43882d003a153/types/lunr/index.d.ts

**Comment by jbouzekri** [comment]
@seantanly : Thank you for the explanation. I was looking at [Microsoft/types-publisher](https://github.com/Microsoft/types-publisher) but I was wondering from where these version numbers were coming from.
